CAIPs/caip-358.md

```diff
@@ -0,0 +1,263 @@
+
+Request example:
+
+```json
+{
+  "version": "1.0.0",
+  "orderId": "order-123456",
+  "acceptedPayments": [
+    {
+      "recipient": "eip155:1:0x71C7656EC7ab88b098defB751B7401B5f6d8976F",
+      "asset": "eip155:1/erc20:0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
+      "amount": "0x5F5E100" 
+    },
+    {
+      "recipient": "solana:4sGjMW1sUnHzSxGspuhpqLDx6wiyjNtZ:9WzDXwBbmkg8ZTbNMqUxvQRAyrZzDsGYdLVL9zYtAWWM",
+      "asset": "solana:4sGjMW1sUnHzSxGspuhpqLDx6wiyjNtZ/slip44:501",
+      "amount": "0x6F05B59D3B20000"
+    }
+  ],
+  "expiry": 1709593200 
+}
+```
+
+#### Response
+
+```typescript
+type PayResult = {
+  version: string;
+  orderId?: string; 
+  txid: string; 
+  recipient: string;  
+  asset: string;  
+  amount: Hex; 
+}
+```
+
+The wallet's response MUST include:
+
+- `txid` with the transaction identifier on the blockchain
+- `recipient` that received the payment. It **MUST** be a valid [CAIP-10][] account ID.
+- `asset` that was used for payment. It **MUST** follow the [CAIP-19][] standard.
+- `amount` that was paid. It **MUST** be represented in hex string
+
+If an `orderId` was provided in the original request, the response **MUST** include the same `orderId`.
+
+`txid` **MUST** be a valid transaction identifier on the blockchain network specified in the asset's chain ID.
+
+ `recipient`, `asset`, and `amount` **MUST** match those specified in the selected direct payment option in the `acceptedPayments` array.
+
+
+Example response:
+```json
+{
+  "version": "1.0.0",
+  "orderId": "order-123456",
+  "txid": "0x8a8c3e0b1b812182db4cabd81c9d6de78e549fa3bf3d505d6e1a2b25a15789ed",
+  "recipient": "eip155:1:0x71C7656EC7ab88b098defB751B7401B5f6d8976F",
+  "asset": "eip155:1/erc20:0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
+  "amount": "0x5F5E100"
+}
+```
+
+#### Idempotency
+
+The `wallet_pay` method **MUST** be idempotent for the same `orderId` when provided. This ensures robustness in case of connection failures or timeout scenarios.
+
+Requirements when `orderId` is provided:
+
+- If a payment with the same `orderId` has already been completed successfully, the wallet **MUST** return the original `PayResult` without executing a new payment
+- If a payment with the same `orderId` is currently pending, the wallet **SHOULD** return the result of the original payment attempt
+- If a payment with the same `orderId` has failed previously, the wallet **MAY** attempt the payment again or return the previous error
+- Wallets **SHOULD** maintain payment status for completed transactions for at least 24 hours after completion
+- If connection is lost during payment execution, dapps **MAY** retry the same request to query the payment status
+
+When `orderId` is not provided:
+
+- Each payment request **SHOULD** be treated as a new payment attempt
+- Wallets **MAY** implement their own deduplication logic based on other request parameters (recipient, asset, amount, expiry)
```

Comment:
```suggestion
- Wallets **MAY**, however, implement their own deduplication logic based on other request parameters (recipient, asset, amount, expiry) to prevent unintended duplicate payments for identical requests
```